### PR TITLE
Implement sample

### LINKED ISSUE #14: Implement sample
Convert the [sample script](https://github.com/broadinstitute/cytominer_scripts/blob/master/sample.R) to an R function.

### PR DESCRIPTION
**Pull request by mcquin**
Resolves #14

### CODE REVIEW

File: R/sample.R [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    plate_list <-
+      suppressMessages(
+        readr::read_csv(
+          paste(metadata_dir, "barcode_platemap.csv", sep = "/"),
+          col_types = readr::cols(Assay_Plate_Barcode = readr::col_character(),
+                                  Plate_Map_Name = readr::col_character()))) %>%
+      dplyr::select("Assay_Plate_Barcode", "Plate_Map_Name") %>%
+      dplyr::inner_join(plate_list_retrieved, by = "Assay_Plate_Barcode") %>%
+      dplyr::group_by("Plate_Map_Name") %>%
+      dplyr::arrange(Assay_Plate_Barcode) %>%
```
Comment (mcquin) [inline R/sample.R]: sample: no visible binding for global variable ‘Assay_Plate_Barcode’

File: R/sample.R [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+      dplyr::mutate(
+        replicate_id = dplyr::dense_rank(.data$Assay_Plate_Barcode)) %>%
```
Comment (mcquin) [inline R/sample.R]: sample: no visible binding for global variable ‘replicate_id’

File: R/sample.R (hunk 1 above)
Comment (mcquin) [reply R/sample.R]: I'm unsure how to resolve this as the documentation states the arguments should be unquoted variable names:

> .data A tbl. All main verbs are S3 generics and provide methods for tbl_df(), dtplyr::tbl_dt() and dbplyr::tbl_dbi().
> ...  **Comma separated list of unquoted variable names.** Use desc() to sort a variable in descending order.
> .by_group If TRUE, will sort first by grouping variable. Applies to grouped data frames only.

File: tests/testthat/test-sample.R [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    workspace_dir = system.file("extdata", package = "cytotools")
+  )
+
+  actual <- readr::read_csv(sampled_csv)
+
+  # Assert we sampled one replicate (one plate)
+  expect_equal(length(unique(actual$Metadata_Plate)), 1)
+
+  # Assert we sampled the whole plate
+  expect_equal(dim(actual)[1], 30)
```
Comment (shntnu) [inline tests/testthat/test-sample.R:46]: Excellent!

**Review by shntnu — APPROVED** [review]
Looks good! I made some formatting changes using styler + added docs. All set.

File: R/sample.R [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+#' @importFrom magrittr %>%
+#' @importFrom magrittr extract2
+#' @importFrom rlang .data
+#' @importFrom utils head
+#' @importFrom utils tail
+#' @export
+sample <- function(batch_id, pattern, output,
+                   replicates = NULL,
+                   workspace_dir = ".") {
+  backend_dir <- paste(workspace_dir, "backend", batch_id, sep = "/")
```
Comment (cells2numbers) [inline R/sample.R]: instead of paste(workspace_dir, "backend", batch_id, sep = "/") use file.path(workspace_dir, "backend", batch_id)

**Comment by mcquin** [comment]
Thanks, @shntnu! ✨

File: R/sample.R [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+  metadata_dir <- paste(workspace_dir, "metadata", batch_id, sep = "/")
```
Comment (cells2numbers) [inline R/sample.R]: file.path() is better suited then paste()

File: R/sample.R [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+          unlist()
+        )
+
+    replicates <- as.integer(replicates)
+
+    # create a plate_list based on number of replicates to be selected
+    plate_list <-
+      suppressMessages(
+        readr::read_csv(
+          paste(metadata_dir, "barcode_platemap.csv", sep = "/"),
```
Comment (cells2numbers) [inline R/sample.R]: replace paste() with file.path()

File: R/sample.R [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        if (length(unlist(lapply(plate_list,
+                                 function(plate) grep(plate, file))))) {
+          file
+        }}) %>%
+      unlist()
+  }
+
+  futile.logger::flog.info(
+    sprintf("Reading %d files...:\n%s", length(file_list),
+            paste(file_list, collapse = "\n"))
```
Comment (cells2numbers) [inline R/sample.R]: replace paste() with file.path()

File: R/sample.R [hunk 8]
```diff
@@ -0,0 +1,89 @@
+utils::globalVariables(c("Assay_Plate_Barcode", "replicate_id"))
+#' Sample selects replicates across specified plates.
+#'
+#' \code{sample} replicates across specified plates, and aggregates the result in a single file.
+#'
+#' @param batch_id        Batch ID.
+#' @param pattern         Regular expression - only CSVs.
+#' @param output          Output file - either CSV or RDS.
```
Comment (cells2numbers) [inline R/sample.R]: Do we also allow feather data format as output?

File: inst/scripts/cytotools_sample [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+Usage:
+  sample -b <id> -f <pattern> [-n <n>] -o <file>
+
+Options:
+  -h --help                             Show this screen.
+  -b <id> --batch_id=<id>               Batch ID.
+  -f <pattern> --pattern=<pattern>      Regular expression - only csv\'s.
+  -n <n> --replicates=<n>               Number of replicates to select per plate map.
+  -o <file> --output=<file>             Output file - either csv or rds.
```
Comment (cells2numbers) [inline inst/scripts/cytotools_sample]: do we allow feather, too?

**Review by cells2numbers — CHANGES_REQUESTED** [review]
Looks good! I agree with Shantanu. Some small changes before we can merge!

File: inst/scripts/cytotools_sample (hunk 9 above)
Comment (shntnu) [reply inst/scripts/cytotools_sample]: Good catch - yes

**Review by shntnu — APPROVED** [review]

**Review by cells2numbers — APPROVED** [review]

**Pull request merged**